File: src/DesireeIA/Enums.cs
// DesireeIA
// Copyright (c) Passaro Francesco Paolo. All rights reserved.
// Licensed under the DesireeIA License - see LICENSE and the "License"
// section of README.md for full terms: no modification, no unauthorized
// integration, no AI training/ingestion without explicit written consent
// from the author.

namespace DesireeIA;

public enum ModelFormat
{
    Unknown = 0,
    Gguf = 1,
    Safetensors = 2
}

public enum InferenceBackend
{
    // 0 = unspecified: the native plan auto-detects it from the hardware
    // (desireeia_make_plan only fills plan.backend when it is 0, see
    // core/plan.cpp/ctx.cpp). ExecutionPlan.Backend uses this as its default
    // on purpose: a partial override like `new ExecutionPlan { ThreadCount =
    // t }` must not silently force Cpu on a machine with a real GPU just
    // because the backend wasn't being set.
    Unconfigured = 0,
    Cpu = 1,
    Cuda = 2,
    Metal = 3,
    Vulkan = 4,
    Intel = 5,
    Axelera = 6
}

public enum Quantization
{
    F32 = 0,
    F16 = 1,
    Q4_0 = 2,
    Q4_1 = 3,
    Q8_0 = 4,
    Q4K = 5,
    Q5K = 6,
    Q6K = 7
}

/// <summary>
/// How <see cref="LocalModel.Predict"/> reuses the K/V cache of the previous call
/// (see <see cref="LocalModel.SessionReuse"/>).
/// </summary>
public enum SessionReuseMode
{
    /// <summary>Always prefill the whole prompt (behavior before 0.0.2).</summary>
    Off = 0,
    /// <summary>Reuse only positions computed by a previous prefill: output identical to a full prefill.</summary>
    Exact = 1,
    /// <summary>Also reuse generated tokens: more reuse, not bit-identical (decode and prefill round differently).</summary>
    IncludeGenerated = 2
}

File: python/desireeia/enums.py
"""DesireeIA enums — 1:1 mapping of C ABI enums from abi.h."""

from enum import IntEnum


class Backend(IntEnum):
    CPU = 1
    CUDA = 2
    METAL = 3
    VULKAN = 4
    INTEL = 5
    AXELERA = 6


class ModelFormat(IntEnum):
    UNKNOWN = 0
    GGUF = 1
    SAFETENSORS = 2


class Quantization(IntEnum):
    F32 = 0
    F16 = 1
    Q4_0 = 2
    Q4_1 = 3
    Q8_0 = 4
    Q4K = 5
    Q5K = 6
    Q6K = 7


class Error(IntEnum):
    OK = 0
    INVALID_ARG = -1
    NOT_SUPPORTED = -2
    IO = -3
    PARSE = -4
    NO_MEM = -5
    UNDEFINED = -6


class SsdTierMode(IntEnum):
    OFF = 0
    AUTO = 1
    ALWAYS = 2


class SpecialToken(IntEnum):
    BOS = 0
    EOS = 1
    UNK = 2
    PAD = 3


class InferenceBackend(IntEnum):
    """Public backend enum. Unconfigured=0 lets the native plan auto-detect."""
    UNCONFIGURED = 0
    CPU = 1
    CUDA = 2
    METAL = 3
    VULKAN = 4
    INTEL = 5
    AXELERA = 6


class StreamType(IntEnum):
    IMAGE = 0
    VIDEO = 1
    FILE = 2
    DOCUMENT = 3
    AUDIO = 4
